### PR TITLE
Refactor SciMLOps after #200

### PR DESCRIPTION
**Pull request by vpuri3**
the changes in #200 introduced `batch` kwarg to functionoperator. The API was that with `batch = true`, the last dimension of `u` will be the batch dim. With `batch = false`, the ND input array will be treated as one "column vector". This caused a lot of size mismatch errors since the rest of SciMLOperators takes the second dimension to be the batch dimension as that is what makes sense for matrices. And cases with ND arrays were handled with reshaping to 2D arrays. The principle issue became that we had different measures of length: `length(u)` for FunctionOperator with `batch = false`, `size(u)[1:end] |> prod` for `FunctionOP` with `batch = true`, and `size(u, 2)` for the rest. For an ND array and an arbitrary wrapper, it would be hard to know which measure to use. An increased reliance on `reshape` also causes confusion that may lead to unsafe behavior. Eg. applying matrices to arrays of wrong sizes should result in errors but could be done with reshape.

With this PR, I am limiting FunctionOperator w. `batch  = true` to only accept 2D arrays where the second dimension is the batch dimension. This is consistent with the rest of the package. FunctionOperator w. `batch = false` is still able to work with ND arrays as before. Further, I am removing operator evaluation and caching methods for ND arrays to avoid confusion. The only operator that is able to work with ND arrays is `FunctionOperator` w. `batch = false`. The rest of the operators work only with `AbstractVecOrMat`s where (optionally) batches are stacked in the 2nd dimension.

### new batch API for FunctionOperator
`batch` - Boolean indicating if the input/output arrays comprise of batched column-vectors stacked in a matrix. If `true`, the input/output arrays must be `AbstractVecOrMat`s, and the length of the second dimension (the batch dimension) must be the same. The batch dimension is not involved in size computation. For example, with `batch = true`, and `size(output), size(input) = (M, K), (N, K)`, the `FunctionOperator` size is set to `(M, N)`. If `batch = false`, which is the default, the `input`/`output` arrays may of any size so long as `ndims(input) == ndims(output)`, and the `size` of `FunctionOperator` is set to `(length(input), length(output))`.

- [x] do size check before mat_mul for FunctionOperator. Other concrete types, ie MatrixOp, ScalarOp manage that already.
- [x] rm `multidim.jl`
- [x] all scalar types must accept `AbstractArray`
- [x] add FunctionOp test inside a tensorprod

This is a major update. Please tag `0.3.0` after merging.

### CODE REVIEW

**Comment by vpuri3** [comment]
looks like rm'ing the redundant functionop constructor didn't break anything. good

File: src/interface.jl [hunk 1]
```diff
@@ -392,49 +348,63 @@ expmv!(v,L::AbstractSciMLOperator,u,p,t) = mul!(v,exp(L,t),u)
...
 end
 
 function Base.:(==)(L1::AbstractSciMLOperator, L2::AbstractSciMLOperator)
+    @warn """using convert-based fallback for Base.=="""
     size(L1) != size(L2) && return false
     convert(AbstractMatrix, L1) == convert(AbstractMatrix, L1)
 end
 
 Base.@propagate_inbounds function Base.getindex(L::AbstractSciMLOperator, I::Vararg{Any,N}) where {N}
+    @warn """using convert-based fallback for Base.getindex"""
```
Comment (vpuri3) [inline src/interface.jl:362]: added warnings to `convert`-based fallbacks.

File: src/matrix.jl [hunk 2]
```diff
@@ -395,27 +395,25 @@ v = L(u, p, t) # == A * u + B * (p * t)
 ```
 
 """
-struct AffineOperator{T,AT,BT,bT,C,F,F!} <: AbstractSciMLOperator{T}
+struct AffineOperator{T,AT,BT,bT,F,F!} <: AbstractSciMLOperator{T}
```
Comment (vpuri3) [inline src/matrix.jl:398]: rewrote operator evaluation methods for affine op to eliminate the need for cache.

File: src/scalar.jl [hunk 3]
```diff
@@ -35,24 +35,24 @@ has_mul!(::AbstractSciMLScalarOperator) = true
 islinear(::AbstractSciMLScalarOperator) = true
 has_adjoint(::AbstractSciMLScalarOperator) = true
 
-Base.:*(α::AbstractSciMLScalarOperator, u::AbstractVecOrMat) = convert(Number, α) * u
```
Comment (vpuri3) [inline src/scalar.jl:38]: since `ScalarOperator` doesn't depend on size, it can be applied to all `AbstractArray`s

File: src/tensor.jl [hunk 4]
```diff
@@ -198,21 +198,21 @@ function cache_internals(L::TensorProductOperator, u::AbstractVecOrMat)
...
 
-    _ , ni = size(inner)
-    _ , no = size(outer)
+    mi , ni = size(inner)
+    _  , no = size(outer)
     k = size(u, 2)
 
     uinner = reshape(u, (ni, no*k))
-    uouter = L.cache[2]
+    uouter = reshape(L.cache[2], (no, mi * k))
```
Comment (vpuri3) [inline src/tensor.jl:206]: array passed to `cache_operator` must be `AbstractVecOrMat` here.

File: src/tensor.jl [hunk 5]
```diff
@@ -529,7 +529,9 @@ function outer_div!(L::TensorProductOperator, u::AbstractVecOrMat)
     U = reshape(U, (ni, no, k))
     C = reshape(C, (no, ni, k))
     permutedims!(C, U, PERM)
+    C = reshape(C, (no, ni * k))
     ldiv!(outer, C)
+    C = reshape(C, (no, ni, k))
     permutedims!(U, C, PERM)
```
Comment (vpuri3) [inline src/tensor.jl:535]: added missing calls to `reshape` for converting 3D array to 2D for mul/div

File: src/utils.jl [hunk 6]
```diff
@@ -1,17 +1,32 @@
 #
-function _mat_sizes(L::AbstractSciMLOperator, u::AbstractArray)
-    m, n = size(L)
-    nk = length(u)
 
-    size_in  = u isa AbstractVecOrMat ? size(u) : (n, nk ÷ n)
-    size_out = issquare(L) ? size_in : (m, size(u)[2:end]...)
+###
```
Comment (vpuri3) [inline src/utils.jl:3]: moved this here from interface.jl

File: src/utils.jl [hunk 7]
```diff
@@ -1,17 +1,32 @@
 #
-function _mat_sizes(L::AbstractSciMLOperator, u::AbstractArray)
```
Comment (vpuri3) [inline src/utils.jl:2]: removed helper functions for `multidim.jl`

File: test/total.jl [hunk 8]
```diff
@@ -80,17 +82,25 @@ end
...
+
     # Introduce update function for B
     B = MatrixOperator(zeros(N,N); update_func! = (A, u, p, t) -> (A .= p))
-    C = rand(N,N)
+
+    # FunctionOp
+    _C = rand(N, N) |> Symmetric
+    f(u, p, t) = _C * u
+    f(v, u, p, t) = mul!(v, _C, u)
+    C = FunctionOperator(f, zeros(N); batch = true, issymmetric = true, p = p)
```
Comment (vpuri3) [inline test/total.jl:98]: test for `FunctionOp` inside a `TensorProdOp`

File: src/func.jl [hunk 9]
```diff
@@ -64,28 +64,6 @@ mutable struct FunctionOperator{iip,oop,mul5,T<:Number,F,Fa,Fi,Fai,Tr,P,Tt,C} <:
     end
 end
 
-function FunctionOperator(op,
```
Comment (vpuri3) [inline src/func.jl:67]: remove obsolete FunctionOp constructor with that used to flatten Nd arrays.

File: src/func.jl [hunk 10]
```diff
@@ -133,7 +111,7 @@ uniform across `op`, `op_adjoint`, `op_inverse`, `op_adjoint_inverse`.
...
-* `batch` - Boolean indicating if the input/output arrays comprise of batched vectors. If `true`, the last dimension of input/output arrays is considered to be the batch dimension and is not involved in size computation. For example, let `size(output), size(input) = (M, K), (N, K)`. If `batch = true`, then the second dimension is assumed to be the batch dimension, and the `size(F::FunctionOperator) = (M, N)`. If `batch = false`, then `size(F::FunctionOperator) = (M * K, M * K)`.
+* `batch` - Boolean indicating if the input/output arrays comprise of batched column-vectors stacked in a matrix. If `true`, the input/output arrays must be `AbstractVecOrMat`s, and the length of the  second dimension (the batch dimension) must be the same. The batch dimension is not involved in size computation. For example, with `batch = true`, and `size(output), size(input) = (M, K), (N, K)`, the `FunctionOperator` size is set to `(M, N)`. If `batch = false`, which is the default, the `input`/`output` arrays may of any size so long as `ndims(input) == ndims(output)`, and the `size` of `FunctionOperator` is set to `(length(input), length(output))`.
```
Comment (vpuri3) [inline src/func.jl:114]: description of change

File: src/func.jl [hunk 11]
```diff
@@ -172,24 +150,56 @@ function FunctionOperator(op,
                           isposdef::Bool = false,
                          ) where{N}
 
+    # establish types
```
Comment (vpuri3) [inline src/func.jl:153]: add comments, error messages to constructor

File: src/func.jl [hunk 12]
```diff
@@ -316,16 +330,63 @@ function iscached(L::FunctionOperator)
     !isnothing(L.cache)
 end
 
-function cache_self(L::FunctionOperator, u::AbstractVecOrMat, v::AbstractVecOrMat)
-    !L.traits.ifcache && @debug """Cache is being allocated for a FunctionOperator
-        created with kwarg ifcache = false."""
-    @set! L.cache = zero.((u, v))
+# fix method amg bw AbstractArray, AbstractVecOrMat
```
Comment (vpuri3) [inline src/func.jl:333]: cache_operator, cache_self for functionop

File: src/func.jl [hunk 13]
```diff
@@ -453,63 +521,135 @@ has_mul!(::FunctionOperator{iip}) where{iip} = iip
 has_ldiv(L::FunctionOperator{iip}) where{iip} = !(L.op_inverse isa Nothing)
 has_ldiv!(L::FunctionOperator{iip}) where{iip} = iip & !(L.op_inverse isa Nothing)
 
+function _sizecheck(L::FunctionOperator, u, v)
```
Comment (vpuri3) [inline src/func.jl:524]: size checks for functionop

**Comment by vpuri3** [comment]
@ChrisRackauckas

File: src/basic.jl [hunk 14]
```diff
@@ -8,7 +8,7 @@ struct IdentityOperator <: AbstractSciMLOperator{Bool}
 end
 
 # constructors
-IdentityOperator(u::AbstractArray) = IdentityOperator(size(u,1))
```
[the remark this answers is not in the document]
Comment (ChrisRackauckas) [reply src/basic.jl:11]: Why is that? IdentityOperator and NullOperator are pretty well defined on the whole cast.

[the remark this answers is not in the document]
Comment (vpuri3) [reply src/basic.jl:11]: I'm removing these constructors altogether. the safest thing would be to pass a length with the `NullOperator(len::Int)` constructor.

**Comment by vpuri3** [comment]
@ChrisRackauckas can you take a second look sometime soon? this is holding some stuff up.

File: src/utils.jl [hunk 15]
```diff
@@ -1,17 +1,32 @@
...
+    _accepted_kwargs = (accepted_kwargs === nothing) ? () : accepted_kwargs 
+    # accepted_kwargs can be passed as nothing to indicate that we should not filter 
+    # (e.g. if the function already accepts all kwargs...). 
+    return (_accepted_kwargs isa NoKwargFilter) ? _update_func : FilterKwargs(_update_func, _accepted_kwargs)
+end
+function update_func_isconstant(update_func)
+    if update_func isa FilterKwargs
+        return update_func.f == DEFAULT_UPDATE_FUNC
+    else
+        return update_func == DEFAULT_UPDATE_FUNC
```
Comment (ChrisRackauckas) [inline src/utils.jl:27]:
```suggestion
        return update_func.f === DEFAULT_UPDATE_FUNC
    else
        return update_func === DEFAULT_UPDATE_FUNC
```

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**